src/Sparrow/Json/PerCoreContainer.cs

```diff
@@ -33,23 +46,34 @@ public bool TryPull(out T output)
 
                 if (Interlocked.CompareExchange(ref coreItems[i], null, cur) != cur)
                     continue;
+
+                Interlocked.Decrement(ref _perCoreArrayLength[currentProcessorId].Value);
                 output = cur;
                 return true;
             }
+
             output = default;
             return false;
         }
 
         public bool TryPush(T cur)
         {
-            var core = _perCoreArrays[CurrentProcessorIdHelper.GetCurrentProcessorId() % _perCoreArrays.Length];
+            int currentProcessorId = CurrentProcessorIdHelper.GetCurrentProcessorId() % _perCoreArrays.Length;
+            if (_perCoreArrayLength[currentProcessorId].Value >= _numberOfSlotsPerCore)
+                return false;
+
+            var core = _perCoreArrays[currentProcessorId];
 
             for (int i = 0; i < core.Length; i++)
             {
                 if (core[i] != null)
                     continue;
+
                 if (Interlocked.CompareExchange(ref core[i], cur, null) == null)
```

Comment:
Sanity check. Every time we do CompareExchange we are either Incrementing or Decrementing as well, those 2 operations obviously are not atomic (both of them together). This is fine, because if we will take concurrency into account, in worst case we will hit prev implementation - a full scan. @grisha-kotler please confirm?

Reply:
That's correct